src/cls/async-hooks.ts

```diff
@@ -64,51 +98,75 @@ export class AsyncHooksCLS<Context extends {}> implements CLS<Context> {
   }
 
   enable(): void {
-    this.currentContext.value = this.defaultContext;
+    this.contexts = {};
     this.hook.enable();
     this.enabled = true;
   }
 
   disable(): void {
-    this.currentContext.value = this.defaultContext;
+    this.contexts = {};
     this.hook.disable();
     this.enabled = false;
   }
 
   getContext(): Context {
-    return this.currentContext.value;
+    // We don't store this.defaultContext directly in this.contexts.
+    // Getting undefined when looking up this.contexts means that it wasn't
+    // set, so return the default context.
+    const current = this.contexts[this.ah.executionAsyncId()];
+    return current ? current.value : this.defaultContext;
   }
 
   setContext(value: Context): void {
-    this.currentContext.value = value;
+    this.contexts[this.ah.executionAsyncId()].value = value;
```

Comment:
`this.contexts[this.ah.executionAsyncId()]` can be undefined if it hasn't been assigned before. How do we make sure that it already exists? Is `setContext()` always called from within `runWithNewContext()`?

Reply:
Good catch, and while the Trace Agent never calls setContext outside of runWithNewContext, it would be for the better that the function doesn't throw an error :P

Reply:
If you intend `setContext()` to be always called from within `runWithNewContext()`, would it be better to assert here?